fedimint-atomic-broadcast/src/lib.rs

```diff
@@ -0,0 +1,246 @@
+//! This module implements fedimints custom atomic broadcast abstraction. A
+//! such, it is responsible for ordering serialized items in the form of byte
+//! vectors. The Broadcast is able to recover from a crash at any time via a
+//! backup that it maintains in the servers [fedimint_core::db::Database]. In
+//! Addition, it stores the history of accepted items in the form of
+//! [SignedBlock]s in the database as well in order to catch up fellow guardians
+//! which have been offline for a prolonged period of time.
+//!
+//! Though the broadcast depends on [fedimint_core] for [fedimint_core::PeerId],
+//! [fedimint_core::encoding::Encodable] and [fedimint_core::db::Database]
+//! it implements no consensus logic specific to Fedimint, to which we will
+//! refer as Fedimint Consensus going forward. To the broadcast a consensus item
+//! is merely a vector of bytes without any further structure.
+//!
+//! # Example Setup
+//!
+//! ```ignore
+//! let block_index = 0;
+//! let (mempool_item_sender, mempool_item_receiver) = async_channel::bounded(256);
+//! let (incoming_message_sender, incoming_message_receiver) = async_channel::bounded(256);
+//! let (outgoing_message_sender, outgoing_message_receiver) = async_channel::bounded(256);
+//! let (ordered_item_sender, ordered_item_receiver) = mpsc::channel(1);
+//! let (shutdown_sender, shutdown_receiver) = watch::channel(None);
+//!
+//! let broadcast_handle = tokio::spawn(fedimint_atomic_broadcast::run(
+//!    keychain,
+//!    db,
+//!    block_index,
+//!    mempool_item_receiver,
+//!    incoming_message_receiver,
+//!    outgoing_message_sender,
+//!    ordered_item_sender,
+//!    shutdown_receiver,
+//! ));
+//! ```
+//!
+//! # The journey of a ConsensusItem
+//!
+//! Let us sketch the journey of an [fedimint_core::epoch::ConsensusItem] into a
+//! signed block.
+//!
+//! * The node which wants to order the item calls consensus_encode to serialize
+//!   it and sends the resulting serialization it to its running atomic
+//!   broadcast instance via the mempool item sender.
+//! * Every 250ms the broadcasts currently running session instance creates a
+//!   new batch from its mempool and attaches it to a unit in the form of a
+//!   UnitData::Batch. The size of a batch and therefore the size of a
+//!   serialization is limited to 10kB.
```

Comment:
Isn't 10kB kind of ... very little? Every 250ms, it's 40KB/s, hardcoded at this very level.

Reply:
its 40KB/s per guardian per second. So just under 10mb per minute in a federation of four. All of which needs to be stored on disk forever if all items are valid. We could probably set this higher, this is conservative for sure, however we would probably be limited by disk space anyways.